projects/ids-enterprise-ng/src/lib/modal-dialog/soho-modal-dialog.ref.ts

```diff
@@ -247,70 +236,91 @@ export class SohoModalDialogRef<T> {
+      })
+
+      this.options({ content });
+    }
   }
 
   /**
-   * Constructor.
+   * Handles destruction of component and modal
    */
-  constructor(private ngZone: NgZone) {
+  private destroy() {
+    if (this.componentRef) {
+      // destroy component
+      this.ngZone.run(() => {
+        this.appRef.detachView(this.componentRef.hostView);
+        this.componentRef.destroy();
+        this.componentRef = null;
+      });
+    }
+
+    if (this.modal) {
+      // destroy modal
+      this.ngZone.runOutsideAngular(() => {
+        this.modal.destroy();
+        this.modal = null;
+      });
+    }
   }
 
+
   /**
    * Opens the dialog.
    *
    * @return the dialog ref.
    */
   open(): SohoModalDialogRef<T> {
-    if (!this.componentRef && !this._options.content) {
-      throw Error('componentRef or content must be initialised.');
-    }
-
-    this.ngZone.runOutsideAngular(() => {
-      // Assume conent ...
-      let element: JQuery = $('body');
-      if (this.componentRef) {
-        // .. unless component supplied, in which case get a selector
-        // to the component and use that.
-        element = jQuery(this.componentRef.location.nativeElement);
-        this._options.content = element;
-      }
-
+    this.jQueryElement = this.ngZone.runOutsideAngular(() => {
+      const element = jQuery('body');
       element.modal(this._options);
-      this.modal = element.data('modal');
-
-      // When the modal is opened, it is moved to the body, so
-      // set the jQueryElement.
-      this.jQueryElement = this.modal.element;
-
-      // Add listeners to control events
-      this.jQueryElement.on('close', ((event: any, isCancelled: boolean) => this.ngZone.run(() => this.onClose(event, isCancelled))));
-      this.jQueryElement.on('afterclose', ((event: any) => this.ngZone.run(() => this.onAfterClose(event))));
-      this.jQueryElement.on('open', ((event: any) => this.ngZone.run(() => this.onOpen(event))));
-      this.jQueryElement.on('afteropen', ((event: any) => this.ngZone.run(() => this.onAfterOpen(event))));
-
-      // These are vetoable events.
-      this.jQueryElement.on('beforeopen', ((event: any) => this.ngZone.run(() => this.onBeforeOpen(event))));
-      this.jQueryElement.on('beforeclose', ((event: any) => this.ngZone.run(() => this.onBeforeClose(event))));
-      this.jQueryElement.on('beforedestroy', ((event: any) => this.ngZone.run(() => this.onBeforeDestroy(event))));
 
+      this.modal = element.data('modal');
+      return this.modal.element;
     });
 
+    // Create Event Observables
+    this.beforeOpen$ = fromEvent(this.jQueryElement, 'beforeopen');
+    this.open$ = fromEvent(this.jQueryElement, 'open');
+    this.afterOpen$ = fromEvent(this.jQueryElement, 'afteropen');
+    this.beforeClose$ = fromEvent(this.jQueryElement, 'beforeclose');
+    this.close$ = fromEvent(this.jQueryElement, 'close');
+    this.afterClose$ = fromEvent(this.jQueryElement, 'afterclose').pipe(
```

Comment:
vetoable events are removed but I have prepared listeners for these events in case they are needed except beforedestroy since it is almost the same with afterclose plus there's also an onDestroy hook by angular so use-case-wise it doesn't make it worth to keep.